### ISSUE TITLE
Question about the evaluation results

### ISSUE THREAD

**Issue by wzever** [ml-jku/DIffUCO#8]
Dear authors, 
Thanks for your insightful work! I have run the testing code following your latest instruction as simple as `python ConditionalExpectation.py --wandb_id m3h9mz5g --dataset RB_iid_small --GPU 0 --evaluation_factor 3 --n_samples 1 --batch_size 1` to evaluate your pretrained model on some new datasets. I appreciate your delicately wrapped codes but I'm somewhat confused by the complicated output of results where many "energy" related terms are reported (`L_energy`, `mean_best_energy`, `mean_normed_energy`, `overall_Loss`, `mean_best_energy_CE`, etc) showing very close values, where I failed to obtain an **exact, hard-guaranteed feasible solution** of the instances, i.e., the decision variables $x\in$ {0, 1} $^N$, for my subsequent study. Therefore, I would very much appreciate it if you can provide some ideas for an accurate and clear measuring of your method. To be specific, I have two questions, 1) could you explain a little the differences of the terms, and which metric should I report as the objective (e.g., set size for MIS, cut size for MaxCut, etc.) for fair comparison in convention to the combinatorial optimization community; and 2) is it convenient to obtain the decision variables within your instructed evaluation pipeline (e.g., which variables in your code after what kind of mathematical transformation)? Btw, I found something like `log_dict["metrics"]["X_0_CE"]` but it seems incorrect when averaged over the test set. Thank you!

**Comment by sanokows** [comment]
Thank you for your interest.

Here are my answers to your questions:

Difference between L_energy, mean_best_energy, mean_normed_energy, overall_Loss, mean_best_energy_CE:
L_energy and overall_Loss are metrics from computing the loss objective of DiffUCO.
In mean_best_energy the best out of n_samples is taken for each CO problem instance and then this is averaged over the test set. 
Please ignore "mean_normed_energy" as it is a relict of old code, so please ignore it.
In mean_best_energy_CE is the same as mean_best_energy but after application of CE.
While mean_best_energy could in principle still contain violations mean_best_energy_CE does not contain violations anymore.
mean_best_energy could in principle still contain violations but it would be easy to remove them with simple postprocessing and this would then correspond to an even lower energy value.
In this case mean_best_energy_CE is minus the set size for MIS. For MaxCut the Maximum Cut Value should be reported as something like MaxCut_Value in wandb.

I was not aware that the solutions "X_0_CE" are averaged before logging. This is of course not correct.
If you want to change this you might do the following:
- retrieve "X_0_CE" in L. 921 of train.py from log_dict and add then into a list
- add this list into eval_log_dict in L. 988. then this method should return the correct solutions.

Are all of your questions adequately addressed?
Feel free to ask further questions if something is unclear.

**Comment by wzever** [comment]
Thanks for your prompt reply. I now understand that the `mean_best_energy_CE` value is supposed to be reported as the objective. But there might be a misunderstanding regarding the `X_0_CE` in my previous description. To my understanding, `X_0_CE` contains the decision variables $x\in$ {0,1} $^N$ for each test problem instance, and if so, the average of the  summation of `X_0_CE`s through the dataset should correspondingly equal `mean_best_energy_CE`. However, I encountered such discrepancy as shown by the following example in the test function in `train.py` (MIS on RB_iid_small, batch_size = 1, n_samples = 1).
```py
objs = []
for iter, (batch_dict) in tqdm(enumerate(dataloader), total=len(dataloader)):
    # ... (model inference)
    objs.append(log_dict_metrics["X_0_CE"].sum().item())
    # ... (other logging process)
print("mean_obj", sum(objs) / len(objs))
```
In the reported results, `test/mean_best_energy_CE` is -19.2, which is quite reasonable, yet the `mean_obj` I defined turned out to be 55.556. So, I wonder if there's anything wrong with my understanding of your codes and all I expect to access is a group of decision variables (as simple as any array-like binary lists in shape of [N,]) that exactly evaluate the same value of `test/mean_best_energy_CE`, via a perhaps tedious summing and averaging for set sizes, cut sizes, clique sizes, etc., by which I respectfully argue that the satisfaction or violation of constraints could also be checked directly with better transparency, especially for those who are not quite familiar with the energy-based metrics for COPs. Thank you so much for your patience!

**Comment by sanokows** [comment]
`test/mean_best_energy_CE` is in your case the averaged energy over the test set right?
Because if the is the energy value of only one CO instance a value of -19.2 would mean that there is something wrong. Each instance should have an integer value after application of CE.

The reason why you get a wrong value after summing up the values of `X_0_CE` is that  `X_0_CE` contains the solution values of an additional padding graph, which is needed to speed up the code by using jax.jit. You can fix the code by removing the solution variables of the padded graph.

You can do this by adapting the code in Trainers/BaseTrainer.py in line 168 with the following code:

  last_node_idx = jnp.sum(energy_graph_batch.n_node[0, 0:-1])
  log_dict["metrics"]["X_0_CE"], log_dict["metrics"]["energies_CE"] = X_0_CE[:, :last_node_idx, ...], energies_CE[:,:-1]
  
 The resulting code should then look like this:
![image](https://github.com/user-attachments/assets/221f5187-0103-4473-90ec-d12a330f6ec7)

Please note that summing up the value of `X_0_CE` is only equal to the energy for some CO problems.
I believe on MaxCut for example this should not work.
When I compute energies in my code I am always using the corresponding energy classes in /EnergyFunctions.

**Comment by wzever** [comment]
Thank you very much for the detailed instruction, which is of great help to me! One minor issue left is when n_samples > 1, the X_0_CE becomes of shape [num_nodes, n_samples, 1]. I tried to retrieve the array corresponding to the argmax index of minus mean_energy or MaxCut_Value_CE but failed to get the correct objective value for the original problem (which works just fine when n_sample set to 1). Could you please check my processing below (in train.py)?
```py
vars = []
for iter, (batch_dict) in tqdm(enumerate(dataloader), total=len(dataloader)):
    # ... (model inference)
    num_nodes, num_samples, _ = log_dict_metrics["X_0_CE"].shape
    best_sample_id = np.argmax(np.abs(batch_CE_log_dict["MaxCut_Value_CE"].reshape(-1))) 
    # batch_CE_log_dict["energy_CE"] for MIS, MaxCl, etc.
    var = log_dict_metrics["X_0_CE"].reshape((num_samples, -1))[best_sample_id].tolist()
    vars.append(var)
    # ... (other logging process)
 
calculate_mean_obj(vars, task="MaxCut") # MaxCl, MIS...
```
Thanks again!

**Comment by sanokows** [comment]
I have checked whether the sum of each state of log_dict_metrics["X_0_CE"] summed up over each solution variable is equal to minus the energy on MIS and this seems to be the case.
In this case, if you do not get the correct solution variables this issue must be related to your code, so If you implement your code correctly I think this must work. 
Unfortunately, I currently do not have the time to look more deeply into your code.

Finally, I would like to note that in case of MaxCut the MaxCut_Value corresponds to the number of cut edges and not the set size of the maximum cut. So summing up log_dict_metrics["X_0_CE"] won't give you a number that equates to the energy in this case.

**Comment by wzever** [comment]
Ok, I will check my code. I certainly know that the objective of MaxCut is the size of cut edges which is different from the others counting the node set and my previous description was simply for an example. Anyway, your response so far has helped me quite a lot. Thanks for your time!

**Comment by sanokows** [comment]
I am happy that I was able to help you.
I will close this issue for now.
If you have any more problems feel free to open another issue.
